keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,556 @@
+
+All those changes have led to a situation when most of the logs produced by distributed applications is now not stored locally but instead is aggregated and sent to some sort of external storage system. This makes working with logs much easier and provides completely new opportunities for using logs for different kinds of quantitative analyzes but at the same time moves the focus for logs from being easy to read for humans to be easy to process and index by logging services.
+
+Because Kubernetes is a very technical solution on which lots of developers work on a daily basis we should not abandon completely developers centered approach for logging but instead add a way for Kubernetes users to choose between those two perspectives depending on the cluster purpose - if the given cluster is used for development or is it a production cluster. However this does not mean that the human-readable output format provided for developers needs to be exactly the same as the current klog format.
+
+Apart from being easy to ingest into logging services logs in order to be useful needs also to be searchable meaning that it should be easy for the user to find the most relevant information in them. Solutions like Elastic or BigQuery makes querying structured logs very easy. What they do not provide is the consistent schema to be used by those logs. Without it querying logs will be still possible but finding specific information might be much harder. Therefore apart from simply having some structure logs produced by applications should also follow a schema which would enforce consistent naming, consistent way of referencing the same objects and common set of metadata to be stored for each of those objects. With such schema in place building queries for logs becomes much easier.
+
+### Goals
+
+* Provide structured logging interface in K8s codebase
+* Standardize commonly used logging keys (e.g. referencing k8s object, http request fields)
+* Propose sound strategy for migration ensuring its completeness
+* Make implementation of context based solutions like logs correlation or tracing within Kubernetes easier.
+
+### Non-Goals
+
+* We are **not** changing logging pipeline in K8s
+* We are **not** standardizing some specific log messages
```

Comment:
10 lines up you talk about "migration ensuring its completeness".  How does that work if you aren't standardizing?

Reply:
So the goal of this effort is not to decide exactly what should be the actual keys and values for very specific log lines like for instance API server http request logs but instead to provide means for this kind of standardization to happen for all logs.

Reply:
We decided to change approach.